st2common/st2common/operators.py

```diff
@@ -105,59 +111,65 @@ def search(value, criteria_pattern, criteria_condition, check_function):
           item.to_value:
             type: "equals"
             pattern: "Approved"
+    ---
+
+    Example #2
+    
+    ---
+    criteria:
+      trigger.fields:
+        type: search
+        # Controls whether this criteria has to match any or all items of the list
+        condition: all2all  # all2any, any2all or any2any
+        pattern:
+          # "#" and text after are ignored. This allows dictionary keys to be unique but refer to the same field
+          # Any text can go after the hash.
+          item.field_name#1:
+            type: "greaterthan"
+            pattern: 40
+
+          item.field_name#2:
+            type: "lessthan"
+            pattern: 50
     """
-    if criteria_condition == "any":
-        # Any item of the list can match all patterns
-        rtn = any(
-            [
-                # Any payload item can match
-                all(
-                    [
-                        # Match all patterns
-                        check_function(
-                            child_criterion_k,
-                            child_criterion_v,
-                            PayloadLookup(
-                                child_payload, prefix=TRIGGER_ITEM_PAYLOAD_PREFIX
-                            ),
-                        )
-                        for child_criterion_k, child_criterion_v in six.iteritems(
-                            criteria_pattern
-                        )
-                    ]
-                )
-                for child_payload in value
-            ]
-        )
-    elif criteria_condition == "all":
-        # Every item of the list must match all patterns
-        rtn = all(
-            [
-                # All payload items must match
-                all(
-                    [
-                        # Match all patterns
-                        check_function(
-                            child_criterion_k,
-                            child_criterion_v,
-                            PayloadLookup(
-                                child_payload, prefix=TRIGGER_ITEM_PAYLOAD_PREFIX
-                            ),
-                        )
-                        for child_criterion_k, child_criterion_v in six.iteritems(
-                            criteria_pattern
-                        )
-                    ]
-                )
-                for child_payload in value
-            ]
-        )
-    else:
+    if isinstance(value, dict):
+        value = [value]
+    criteria_condition_list = criteria_condition.split('2', 1)
+    if (not((len(criteria_condition_list) == 1 and (criteria_condition_list[0] == 'any' or criteria_condition_list[0] == 'all')) or
+       (len(criteria_condition_list) == 2 and (criteria_condition_list[0] == 'any' or criteria_condition_list[0] == 'all') and
+       (criteria_condition_list[1] == 'any' or criteria_condition_list[1] == 'all')))):
```

Comment:
Please avoid trying to split `any2any` here. This is a confusing construct. 

Use the explicit whitelist instead.